jobsdb/jobsdb.go

```diff
@@ -2341,13 +2347,24 @@ func (jd *HandleT) getProcessedJobsDS(ds dataSetT, getAll bool, limitCount int,
 			ds.JobTable, ds.JobStatusTable, stateQuery, customValQuery, sourceQuery, limitQuery)
 
 		args := []interface{}{getTimeNowFunc()}
+
+		var wrapQuery []string
 		if params.EventCount > 0 {
-			sqlStatement = fmt.Sprintf(`SELECT * FROM (`+sqlStatement+`) t WHERE running_event_counts - t.event_count + 1 <= $%d;`, len(args)+1)
 			// EXPLAIN `running_event_counts - t.event_count + 1`: If the event count limit "splits" a job we want this jobs to be returned.
 			//				`+1` prevents a job with event count of 1 to be returned.
```

Comment:
I fail to understand the purpose of `+1`, let alone if this is indeed required in the `payload_size` query

Reply:
I am responsible for this ambiguity, so let me try to explain:

The reason behind this: `running_event_counts - t.event_count + 1` is to deal with situations where a single item that the pickup size. For example, the max size is 10_000 bytes and we have an item with 11_000 bytes.

We have two options: pick or not pick the overflowing item. Because we wanted to play it safe, we prefer to pick an oversized item rather than pick nothing at all. Not being able to pick an item, due to an incorrect config or unexpected job size could lead to the processor being stuck.

We prefer availability, this way under any conditions, we ensure to pick at least one item. Rather than correctness: always return 

This is what `running_event_counts - t.event_count` is doing.

Now for the `+1`, if we used just `running_event_counts - t.event_count` we would always pick one more item even if the item size is 1 (which is not possible to overflow). 

The ideal way to handle this is to only allow overflow when it will otherwise lead to not picking a large item and getting stuck.
`running_event_counts - t.event_count + 1` was an easy way to compromise, but it hinders readability.

Reply:
For payload_size that `1` is not common/possible the `+1` can be removed.

Reply:
Agreed. In hindsight, it shouldn't have made it in.

Reply:
> running_event_counts <= $%d OR ROW_NUMBER = 1

say ds-1 has jobs with event_count 1,2,3,10 and ds-2 has jobs with event-count 4
in this case first three jobs from ds-1 and the first job from ds-2 will be picked, leaving 4th job in ds-1 behind.
this would lead to an event ordering issue..?

Reply:
You are right, but still this is something that can be solved if getProcessedJobsDS returns an additional value indicating whether the limit has been reached or not, so that the caller knows when to stop and not rely on the result's size